Fix part of #2132: Chapter List Page Issues

viktoriias opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->
@anandwana001 PTAL

## Explanation
Fixes part of #2132
- Elevate the inactive cards
- Active cards show full description.
- Inactive cards should show Chapter name and description show prerequisites to unlock it.
- Blur thumbnails of inactive cards

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

viktoriias committed:
Fix part of #2132

anandwana001 commented on app/src/main/res/layout/story_chapter_view.xml:
```diff
@@ -91,8 +92,7 @@
             android:fontFamily="sans-serif"
-            android:maxLines="4"
-            android:text="@{htmlContent}"
+            android:text="@{viewModel.chapterSummary.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? htmlContent : String.format(@string/chapter_prerequisite, viewModel.index)}"
```
We also need to show chapter name and it's in Roboto Medium, #555555

<img width="328" alt="Screenshot 2020-12-07 at 08 55 21" src="https://user-images.githubusercontent.com/25057618/101306057-f64e9a00-3869-11eb-9dcc-25eb37c47c7a.png">

<img width="296" alt="Screenshot 2020-12-07 at 08 57 41" src="https://user-images.githubusercontent.com/25057618/101306214-4af21500-386a-11eb-80ab-22c22f88bf8a.png">

anandwana001 replied:
https://github.com/oppia/oppia-android/blob/f616e8570ee6c791250a0675e0b728204b96b1d1/app/src/main/java/org/oppia/android/app/story/StoryViewModel.kt#L102

1. We can check how we can get the name of the previous chapter because of which any chapter is locked and access it and make it a part of the ViewModel (`StoryChapterSummaryViewModel`) item which we are using to show a locked item. It's like introducing a new parameter `previousChapter`
2. If there is any other way you can think which best fit here on how to access the previous chapter name, we can use it too.

viktoriias replied:
Done. Added a new parameter `previousChapterTitle`.
Regarding Roboto Medium, #555555 - is it for the whole prerequisite? <b>Chapter title</b> is being ignored, not sure why?

viktoriias replied:
Html tags in strings.xml don't work with `android:text`. What's the preferred alternative? @anandwana001

anandwana001 replied:
It looks like this is some edge case where styling the passed string need some extra things. 
Please have a look here - https://developer.android.com/guide/topics/resources/string-resource

viktoriias replied:
I see that I need to use `Html.fromHtml`, but I don't think I can use it in xml file, can I?

anandwana001 replied:
I don't think so, as per our discussion in the meeting, we can do it kotlin side.

anandwana001 replied:
Resolving this as we can skip this part from this PR 
reference - https://github.com/oppia/oppia-android/pull/2232#discussion_r541472919

anandwana001 commented on app/src/main/res/layout/story_chapter_view.xml:
```diff
@@ -42,6 +42,7 @@
 
         <org.oppia.android.app.customview.LessonThumbnailImageView
           android:id="@+id/chapter_thumbnail"
+          android:alpha="@{viewModel.chapterSummary.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? 1.0f : 0.3f}"
```
Nit formatting, you can also use Reformat code from the menu in the toolbar or double shift to open quick search in android studio to find reformat code option. 

```xml
 <org.oppia.android.app.customview.LessonThumbnailImageView
          android:id="@+id/chapter_thumbnail"
          android:layout_width="match_parent"
          android:layout_height="0dp"
          android:alpha="@{viewModel.chapterSummary.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? 1.0f : 0.3f}"
          android:scaleType="centerInside"
          app:entityId="@{viewModel.storyId}"
          app:entityType="@{viewModel.entityType}"
          app:layout_constraintDimensionRatio="16:9"
          app:layout_constraintEnd_toEndOf="parent"
          app:layout_constraintStart_toStartOf="parent"
          app:layout_constraintTop_toTopOf="parent"
          app:lessonThumbnail="@{viewModel.chapterThumbnail}" />
```

viktoriias replied:
I ran Reformat code, but I don't think anything changed.

anandwana001 replied:
It will rearrange the lines above.

viktoriias replied:
Ah, to arrange alphabetically? I removed `android:alpha` line, so it should be ok now.

anandwana001 replied:
It's like arrange things alphabetically, spacing, empty line removals, all this will be done automatically with the reformat code option. 

Also, I suggest if you could set up macros, it will help you a lot through the project for all the files. 
Check the wiki guide here - https://github.com/oppia/oppia-android/wiki/Ktlint-Guide#macros

![Screenshot 2020-12-12 at 09 18 54](https://user-images.githubusercontent.com/25057618/101973045-13fd7400-3c5b-11eb-8b1e-530aeb908b6c.png)

viktoriias replied:
Thanks.

anandwana001 commented on app/src/main/res/layout/story_chapter_view.xml:
```diff
@@ -42,6 +42,7 @@
 
         <org.oppia.android.app.customview.LessonThumbnailImageView
```
There are files related to other configuration for this file, We need the same implementation there too. 
`layout-land/story_chapter_view.xml`
`layout-sw600dp/story_chapter_view.xml`

viktoriias replied:
Done.

anandwana001 commented on app/src/main/res/values/strings.xml:
```diff
@@ -63,6 +63,7 @@
   <string name="topic_play_chapter_index">%s.</string>
   <string name="chapter_name">Chapter %s: %s</string>
   <string name="chapter_completed">Chapter %s: %s is completed.</string>
+  <string name="chapter_prerequisite">Complete Chapter %s to unlock this chapter.</string>
```
Added screenshot above for string formatting.

anandwana001 reviewed (changes requested):
Thanks, @viktoriias for your first Pull Request. 

Added a few changes, PTAL

prayutsu commented:
@viktoriias Can you please update the PR title to follow our conventional format?
e.g. - ```Fix part of #2132: Chapter List Page Issues```

viktoriias committed:
Fix part of #2132

viktoriias committed:
Fix part of #2132

anandwana001 commented:
@viktoriias no changes required in `.idea/codeStyles/Project.xml` file.

viktoriias committed:
Fix part of #2132

viktoriias committed:
Fix part of #2132

viktoriias commented:
> @viktoriias no changes required in `.idea/codeStyles/Project.xml` file.
How do I remove this file from here?

anandwana001 commented:
> > @viktoriias no changes required in `.idea/codeStyles/Project.xml` file.
> > How do I remove this file from here?

https://stackoverflow.com/questions/36763274/how-to-remove-a-file-from-git-pull-request/42405903
This might help, removing an unwanted file from a pull request.

viktoriias committed:
Fix part of #2132

viktoriias committed:
Fix part of #2132

viktoriias commented:
> > > @viktoriias no changes required in `.idea/codeStyles/Project.xml` file.
> > > How do I remove this file from here?
> 
> https://stackoverflow.com/questions/36763274/how-to-remove-a-file-from-git-pull-request/42405903
> This might help, removing an unwanted file from a pull request.

Ok thanks, I think I removed it.

BenHenning commented on app/src/main/java/org/oppia/android/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt:
```diff
@@ -22,6 +22,7 @@ class StoryChapterSummaryViewModel(
   val topicId: String,
   val storyId: String,
   val chapterSummary: ChapterSummary,
+  val previousChapterTitle: String,
```
While it's true that currently the previous chapter is always blocking the next one, I think it would be a bit cleaner if we pass along the computed blocking chapter summary (as a ``ChapterSummary`` object) & reference its name in XML in the same way we do for the summary of this card.

I think to do that we will need to add a 'prerequisite_chapter' field to the summary proto: https://github.com/oppia/oppia-android/blob/74db288b9ac25fb942a8feb35e918bcd7e94e55c/model/src/main/proto/topic.proto#L86 & pipe it from the domain controller (https://github.com/oppia/oppia-android/blob/2c02e06d768d5341569e326d9ec042c91c3586da/domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt).

This also solves another part I'm concerned with: it provides a well-defined default if there is no blocking chapter (default ChapterSummary proto).

Note that ideally we would model this as a oneof in ChapterSummary, but I think that could be filed as a future issue for someone else to take on since it will require a broader refactoring.

viktoriias replied:
I don't think I know how to "pipe it from the domain controller"

viktoriias replied:
PTAL, if that's what you had in mind before I add tests.

anandwana001 replied:
Tried to debug the app and It seems like `prerequisiteChapter` is always containing the null items. Also, I tried running your PR, The name of the previous chapter is not visible. 

I will look into this in detail later today my time. 

![Screenshot 2020-12-11 at 09 39 52](https://user-images.githubusercontent.com/25057618/101861649-8bb59b00-3b96-11eb-8bff-2d9af173d160.png)

viktoriias replied:
`missingPrerequisiteChapter` is only present in locked chapters. So I guess, in some cases it's intended.

viktoriias replied:
I added setting `missingPrerequisiteChapter` into one more function,I think that was the problem. PTAL

BenHenning replied:
Ah, nice catch. @viktoriias can you double check that the new tests you added in TopicListController cover both functions? A quick check could be to comment out each of the changes in TopicListController one-at-a-time to verify that one or more of the tests changes fail.

viktoriias replied:
Tests were failing, that's why I had to look for another place to initialize the `missingPrerequisiteChapter`.

> > > Thanks @viktoriias. Overall, this looks great! I have one nit, and one main comment: you added a test for TopicControllerTest, but can you also add a UI-facing test to make sure the binding logic in story_chapter_view is tested? I think that would require a new test in [StoryFragmentTest](https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt). Please add a test for both portrait & landscape (the latter can be done by changing orientation--see `testStoryFragment_changeConfiguration_chapterSummaryIsShownCorrectly`).
> > 
> > 
> > I looked into this test file initially, per @anandwana001's advice. It marks first story as completed, so prerequisite didn't show up. I'll need to remove `markPartialStoryProgressForFractions` to be able to test summary for the second story.
> 
> Removing `markPartialStoryProgressForFractions ` might cause failure for other test. I think what we can do here is to shift it from the `@Before` to all the required test before launching the test. This way we can have test specific weather or not we require partial progress.

I only affects `testStoryFragment_correctStoryCountLoadedInHeader` test: `1 of 2 Chapters Completed` -> `0 of 2 Chapters Completed`. I think it's fine to remove it, let me know what you think.

viktoriias replied:
The tests I added were failing at first, so I had to look for another place to initialize `missingPrerequisiteChapter`.

viktoriias replied:
Tests in TopicListController don't fully cover `combineTopicAndTopicProgress` and `combineStorySummaryAndStoryProgress`. In particular, they don't cover my original change, I only got visual confirmation when running the app. I'm not sure how to get to that particular scenario in `combineStorySummaryAndStoryProgress`..

BenHenning replied:
Good catch. I think we should go ahead and add a new test to cover the other branch. I think the branch that's being missed is indicating that our tests aren't actually completing a story before retrieving the chapter list. Could we adapt the test that fails when the other branch is commented out to cover this case?

viktoriias replied:
Done. I added one more test (and verified it can fail).

BenHenning replied:
Awesome--the new test looks good!

BenHenning commented on app/src/main/res/layout/story_chapter_view.xml:
```diff
@@ -89,10 +90,8 @@
             android:fontFamily="sans-serif"
-            android:maxLines="4"
-            android:text="@{htmlContent}"
+            android:text="@{viewModel.chapterSummary.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? htmlContent : String.format(@string/chapter_prerequisite, viewModel.index, viewModel.previousChapterTitle)}"
```
Databinding has built-in formatting, so this can be simplified to:

```
@{viewModel.chapterSummary.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? htmlContent : @string/chapter_prerequisite(viewModel.index, viewModel.previousChapterTitle)}
```

viktoriias replied:
Nice, thanks!

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -63,6 +63,7 @@
   <string name="topic_play_chapter_index">%s.</string>
   <string name="chapter_name">Chapter %s: %s</string>
   <string name="chapter_completed">Chapter %s: %s is completed.</string>
+  <string name="chapter_prerequisite">Complete Chapter <b>%s: %s</b> to unlock this chapter.</string>
```
Some of these strings need some work to be made more clear, but I think we can add more context for this string in particular. Perhaps: ``chapter_prerequisite_title_label`` might be clearer. I think my issue with the current label is it's a bit vague--we want to try to contextualize it to this view. However, I'm having slight difficulties at the moment coming up with a clearer name. Any thoughts?

viktoriias replied:
Renamed to `chapter_prerequisite_title_label`.

BenHenning reviewed (changes requested):
Thanks @viktoriias! I think this is a great start. I have a few comments, one of which will required changes to part of the overall approach.

Beyond that, can you also add a test to verify that the previous chapter text is correctly computed & used in the expected cases?

viktoriias committed:
Pass a ChapterSummary object as a prerequisite to unlock this chapter.
Remove blurring thumbnails of inactive cards - this change will be done separately.

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -88,6 +88,9 @@ message ChapterSummary {
   // The thumbnail corresponding to this chapter.
   LessonThumbnail chapter_thumbnail = 5;
+
+  // Prerequisite chapter to unlock this chapter.
```
Suggest adding a note here that this is only defined when the play state is ``NOT_PLAYABLE_MISSING_PREREQUISITES``.

I suggest also calling it ``missing_prerequisite_chapter`` to make this more clear in code, as well. There's technically always a prereq chapter except for the first in a story, but we really want to emphasize this is a case where the prerequisite is missing.

viktoriias replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -333,24 +333,21 @@ class TopicController @Inject constructor(
-            val chapterBuilder = chapterSummary.toBuilder()
             chapterBuilder.chapterPlayState = ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES
-            storyBuilder.setChapter(chapterIndex, chapterBuilder)
+            chapterBuilder.prerequisiteChapter = storyBuilder.getChapter(chapterIndex - 1)
```
Nit: could extract the previous chapter as a val at the beginning of the outer else{} since it's used in the if-check above (& this allows for some simple reformatting to make the code a bit easier to read).

viktoriias replied:
Done.

BenHenning commented on app/src/main/res/layout-land/story_chapter_view.xml:
```diff
@@ -77,7 +77,7 @@
             android:layout_marginBottom="4dp"
             android:fontFamily="sans-serif"
-            android:text="@{viewModel.chapterSummary.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? htmlContent : String.format(@string/chapter_prerequisite, viewModel.index, viewModel.previousChapterTitle)}"
+            android:text="@{viewModel.chapterSummary.chapterPlayState != ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES ? htmlContent : @string/chapter_prerequisite_title_label(viewModel.index, viewModel.prerequisiteChapter.name)}"
```
Per discussion in the team meeting, I believe we need to use HtmlParser in order to properly render the bolded text. Unfortunately, it's a bit obscure how to use it. It will actually significant complicate this PR, so what I suggest instead is:
1. Remove the bolded tags from the string for now (since we don't want them to show up if we can't render them)
2. @anandwana001 can you file an issue to track adding rich-text support for this view? It's a non-starter issue since it requires someone to add a custom view somewhere in this recycler view hierarchy & hook it up to the ViewComponent in order to support injecting HtmlParser and using it. Or, we can wait until #2013 is done, but we still need an issue to track making this new title string what it's supposed to be.

anandwana001 replied:
Sounds good to me.

viktoriias replied:
Done. I removed the tags.

anandwana001 replied:
https://github.com/oppia/oppia-android/issues/2252

BenHenning replied:
Thanks!

BenHenning reviewed (changes requested):
Thanks @viktoriias! As a reminder, please also add a test to verify these changes. And, can you please take a look at @anandwana001's reply to my comment thread above about the data potentially being empty in some cases? (I think this might be intentional & correct, but I haven't looked the case closely).

viktoriias committed:
Added tests.

viktoriias committed:
Added tests.

viktoriias committed:
Set missingPrerequisiteChapter in setFirstChapterAsNotStarted function.

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -89,8 +89,8 @@ message ChapterSummary {
 
-  // Prerequisite chapter to unlock this chapter.
-  ChapterSummary prerequisite_chapter = 6;
+  // Prerequisite chapter to unlock this chapter. It is only defined when the play state of this chapter is NOT_PLAYABLE_MISSING_PREREQUISITES.
```
Nit: please line-wrap at 100 characters.

viktoriias replied:
Done.

BenHenning reviewed (changes requested):
Thanks @viktoriias. Overall, this looks great! I have one nit, and one main comment: you added a test for TopicControllerTest, but can you also add a UI-facing test to make sure the binding logic in story_chapter_view is tested? I think that would require a new test in [StoryFragmentTest](https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt). Please add a test for both portrait & landscape (the latter can be done by changing orientation--see ``testStoryFragment_changeConfiguration_chapterSummaryIsShownCorrectly``).

viktoriias commented:
> Thanks @viktoriias. Overall, this looks great! I have one nit, and one main comment: you added a test for TopicControllerTest, but can you also add a UI-facing test to make sure the binding logic in story_chapter_view is tested? I think that would require a new test in [StoryFragmentTest](https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt). Please add a test for both portrait & landscape (the latter can be done by changing orientation--see `testStoryFragment_changeConfiguration_chapterSummaryIsShownCorrectly`).

I looked into this test file initially, per @anandwana001's advice. It marks first story as completed, so prerequisite didn't show up. I'll need to remove `markPartialStoryProgressForFractions` to be able to test summary for the second story.

anandwana001 commented:
> > Thanks @viktoriias. Overall, this looks great! I have one nit, and one main comment: you added a test for TopicControllerTest, but can you also add a UI-facing test to make sure the binding logic in story_chapter_view is tested? I think that would require a new test in [StoryFragmentTest](https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt). Please add a test for both portrait & landscape (the latter can be done by changing orientation--see `testStoryFragment_changeConfiguration_chapterSummaryIsShownCorrectly`).
> 
> I looked into this test file initially, per @anandwana001's advice. It marks first story as completed, so prerequisite didn't show up. I'll need to remove `markPartialStoryProgressForFractions` to be able to test summary for the second story.

Removing `markPartialStoryProgressForFractions ` might cause failure for other test. I think what we can do here is to shift it from the `@Before` to all the required test before launching the test. This way we can have test specific weather or not we require partial progress.

viktoriias commented:
> > > Thanks @viktoriias. Overall, this looks great! I have one nit, and one main comment: you added a test for TopicControllerTest, but can you also add a UI-facing test to make sure the binding logic in story_chapter_view is tested? I think that would require a new test in [StoryFragmentTest](https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt). Please add a test for both portrait & landscape (the latter can be done by changing orientation--see `testStoryFragment_changeConfiguration_chapterSummaryIsShownCorrectly`).
> > 
> > 
> > I looked into this test file initially, per @anandwana001's advice. It marks first story as completed, so prerequisite didn't show up. I'll need to remove `markPartialStoryProgressForFractions` to be able to test summary for the second story.
> 
> Removing `markPartialStoryProgressForFractions ` might cause failure for other test. I think what we can do here is to shift it from the `@Before` to all the required test before launching the test. This way we can have test specific weather or not we require partial progress.

It only affects `testStoryFragment_correctStoryCountLoadedInHeader` test: `1 of 2 Chapters Completed` -> `0 of 2 Chapters Completed`. I think it's fine to remove it, let me know what you think.

anandwana001 commented:
> > > > Thanks @viktoriias. Overall, this looks great! I have one nit, and one main comment: you added a test for TopicControllerTest, but can you also add a UI-facing test to make sure the binding logic in story_chapter_view is tested? I think that would require a new test in [StoryFragmentTest](https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt). Please add a test for both portrait & landscape (the latter can be done by changing orientation--see `testStoryFragment_changeConfiguration_chapterSummaryIsShownCorrectly`).
> > > 
> > > 
> > > I looked into this test file initially, per @anandwana001's advice. It marks first story as completed, so prerequisite didn't show up. I'll need to remove `markPartialStoryProgressForFractions` to be able to test summary for the second story.
> > 
> > 
> > Removing `markPartialStoryProgressForFractions ` might cause failure for other test. I think what we can do here is to shift it from the `@Before` to all the required test before launching the test. This way we can have test specific weather or not we require partial progress.

> 
> It only affects `testStoryFragment_correctStoryCountLoadedInHeader` test: `1 of 2 Chapters Completed` -> `0 of 2 Chapters Completed`. I think it's fine to remove it, let me know what you think.

Sounds good, we can shift it to the test before the launch and check if all the tests are passing on espresso and robolectric.

anandwana001 commented:
> > > > > Thanks @viktoriias. Overall, this looks great! I have one nit, and one main comment: you added a test for TopicControllerTest, but can you also add a UI-facing test to make sure the binding logic in story_chapter_view is tested? I think that would require a new test in [StoryFragmentTest](https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt). Please add a test for both portrait & landscape (the latter can be done by changing orientation--see `testStoryFragment_changeConfiguration_chapterSummaryIsShownCorrectly`).
> > > > 
> > > > 
> > > > I looked into this test file initially, per @anandwana001's advice. It marks first story as completed, so prerequisite didn't show up. I'll need to remove `markPartialStoryProgressForFractions` to be able to test summary for the second story.
> > > 
> > > 
> > > Removing `markPartialStoryProgressForFractions ` might cause failure for other test. I think what we can do here is to shift it from the `@Before` to all the required test before launching the test. This way we can have test specific weather or not we require partial progress.
> 
> > It only affects `testStoryFragment_correctStoryCountLoadedInHeader` test: `1 of 2 Chapters Completed` -> `0 of 2 Chapters Completed`. I think it's fine to remove it, let me know what you think.
> 
> Sounds good, we can shift it to the test before the launch and check if all the tests are passing on espresso and robolectric.

viktoriias committed:
Add a UI-facing test.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt:
```diff
@@ -159,6 +155,10 @@ class StoryFragmentTest {
   @Test
   fun testStoryFragment_correctStoryCountLoadedInHeader() {
     launch<StoryActivity>(createStoryActivityIntent()).use {
+      storyProgressTestHelper.markPartialStoryProgressForFractions(
```
Is this test case is passing on both the espresso and robolectric?

viktoriias replied:
It passed when I ran it. I made an intentional typo to see it fail, and it did.

viktoriias committed:
Add more test coverage for a completed chapter.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt:
```diff
@@ -242,6 +242,41 @@ class StoryFragmentTest {
+  }
+
+  @Test
+  fun testStoryFragment_chapterMissingPrerequisiteIsShownCorrectly() {
```
Small nit, but we usually put the orientation changed test after the non-orientation changed test since it's "building upon" previous assumptions verified by earlier tests.

viktoriias replied:
Ok, I moved it.

BenHenning reviewed (approved):
Thanks @viktoriias! Overall, the PR LGTM. Just have one more comment from me to address.

viktoriias committed:
Move landscape test below the portrait one.

viktoriias commented:
@anandwana001 please review

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt:
```diff
@@ -159,6 +155,10 @@ class StoryFragmentTest {
   @Test
   fun testStoryFragment_correctStoryCountLoadedInHeader() {
     launch<StoryActivity>(createStoryActivityIntent()).use {
+      storyProgressTestHelper.markPartialStoryProgressForFractions(
```
I suggest creating a helper function, might help in future as well to prevent the repetition of this code, and call it before launching the screen as it is the data modification which we do before the screen gets launched. 

```kotlin
private fun setStoryPartialProgress() {}
```

```kotlin
@Test
  fun testStoryFragment_correctStoryCountLoadedInHeader() {
    setStoryPartialProgress
    launch<StoryActivity>(createStoryActivityIntent()).use {....}
    ....
}
```

for other tests too.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt:
```diff
@@ -242,6 +242,41 @@ class StoryFragmentTest {
+          2
+        )
+      )
+      onView(atPositionOnView(R.id.story_chapter_list, 2, R.id.chapter_summary)).check(
```
nit formatting

```kotlin
      onView(
        atPositionOnView(
          R.id.story_chapter_list,
          2,
          R.id.chapter_summary
        )
      ).check(
        matches(
          withText("Complete Chapter 1: What is a Fraction? to unlock this chapter.")
        )
      )
```

apply for other tests too.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt:
```diff
@@ -242,6 +242,41 @@ class StoryFragmentTest {
+  }
+
+  @Test
+  fun testStoryFragment_changeConfiguration_chapterMissingPrerequisiteIsShownCorrectly() {
```
We also need test cases which tests:
- `Active cards show full description.`

Earlier it shows only 4 lines, now we are showing full description, so we a test cases which test this point as well. 

For this, you will need a particular story to check the description. In this test, we are using `createStoryActivityIntent` by opens the `FRACTIONS_STORY_ID_0` which I don't think has a longer description to test. 

@rt4914 Can we update any of the `test_story_id_*.json` where we can have `"outline"` with a longer description to test if the description is visible completely or not in the story card in story activity. 
Or to prevent depending on the test JSON file, can we modify any of the topics JSON file?

rt4914 replied:
@anandwana001 Yes, we can update the `test_story_id_1/2.json`.
While making that change please do following things:
1. Run all **domain** module test and if a test fails because of this change, update the test case accordingly.
2. Run all **app** module tests and if a test fails because of this change, update the test case accordingly.

viktoriias replied:
Done,  **domain** and **app** tests pass.

anandwana001 reviewed (changes requested):
Reviewed the UI test cases, Thanks @viktoriias

viktoriias committed:
Added long summary test + reformatting.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt:
```diff
@@ -135,6 +137,22 @@ class StoryFragmentTest {
+    )
+  }
+
+  private fun setStoryPartialProgressForFractions() {
```
Can we shift this at the end of all the test, above `setUpTestApplicationComponent`, as it is a helper function.

anandwana001 reviewed (approved):
last nit change. Fix ktlint issues.
LGTM, Thanks @viktoriias

Also, just for the future reference, please add 2 screenshots showing that all the tests are passing on Espresso and Robolectric.

viktoriias committed:
Formatting.

viktoriias committed:
Lint

viktoriias commented:
Passing tests:
<img width="1391" alt="Screen Shot 2020-12-16 at 11 27 57 AM" src="https://user-images.githubusercontent.com/10843277/102411614-10455680-3fa7-11eb-8a9d-e70e01c6912c.png">

anandwana001 commented:
Adding @BenHenning for final merge.

viktoriias commented:
StoryFragmentTest Robolectric:
<img width="1372" alt="Screen Shot 2020-12-16 at 5 52 55 PM" src="https://user-images.githubusercontent.com/10843277/102430603-b739ea80-3fc7-11eb-83dd-730b2ebdd07d.png">

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt:
```diff
@@ -330,6 +402,31 @@ class StoryFragmentTest {
     }
   }
 
+  private fun createStoryActivityIntent(): Intent {
```
Prefer: ``createFractionsStoryActivityIntent`` to contrast from the test one below. Or possibly better: have a single 'createStoryActivityIntent' that takes the topic & story IDs to provide exact context at the callsite.

viktoriias replied:
There's no point in having a `createStoryActivityIntent` as a separate function - all it does is call another function. I think it was created to remove code duplication, which we don't want to re-introduce. I like the new name - `createFractionsStoryActivityIntent `, I'll go with that option.

anandwana001 replied:
resolving on behalf of Ben.

BenHenning reviewed (approved):
Thanks @viktoriias! This looks great! Just one small comment, but otherwise LGTM.

@rt4914 or @anandwana001 feel free to merge this after the last comment is resolved if I miss it.

viktoriias committed:
Renaming
